### PR TITLE
SONARJAVA-2256 Cover equals and Objects.equals for S1764

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/IdenticalOperandOnBinaryExpressionCheck.java [hunk 1]
```diff
@@ -69,18 +79,46 @@
       .add(Tree.Kind.OR)
       .add(Tree.Kind.CONDITIONAL_AND)
       .add(Tree.Kind.CONDITIONAL_OR)
+      .add(Tree.Kind.METHOD_INVOCATION)
       .build();
   }
 
   @Override
   public void visitNode(Tree tree) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/IdenticalOperandOnBinaryExpressionCheck.java]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![CRITICAL][CRITICAL] 1 critical

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'

File: java-checks/src/main/java/org/sonar/java/checks/IdenticalOperandOnBinaryExpressionCheck.java [hunk 2]
```diff
@@ -36,6 +39,13 @@
 @Rule(key = "S1764")
 public class IdenticalOperandOnBinaryExpressionCheck extends IssuableSubscriptionVisitor {
 
+  private static final String JAVA_LANG_OBJECT = "java.lang.Object";
+  private static final MethodMatcher EQUALS_MATCHER = MethodMatcher.create().typeDefinition(JAVA_LANG_OBJECT).name("equals").addParameter(JAVA_LANG_OBJECT);
+  private static final MethodMatcher DEEP_EQUALS_MATCHER = MethodMatcher.create().typeDefinition("java.util.Objects")
+    .name("equals").addParameter("java.lang.Object").addParameter(JAVA_LANG_OBJECT);
```
Comment (saberduck) [inline java-checks/src/main/java/org/sonar/java/checks/IdenticalOperandOnBinaryExpressionCheck.java]: `JAVA_LANG_OBJECT` constant should be used for first call to `addParameter`

File: java-checks/src/main/java/org/sonar/java/checks/IdenticalOperandOnBinaryExpressionCheck.java [hunk 3]
```diff
@@ -69,18 +79,46 @@
...
       .add(Tree.Kind.CONDITIONAL_AND)
       .add(Tree.Kind.CONDITIONAL_OR)
+      .add(Tree.Kind.METHOD_INVOCATION)
       .build();
   }
 
   @Override
   public void visitNode(Tree tree) {
+    if (tree.is(Tree.Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree mit = (MethodInvocationTree) tree;
```
Comment (saberduck) [inline java-checks/src/main/java/org/sonar/java/checks/IdenticalOperandOnBinaryExpressionCheck.java]: Perhaps consider extracting this into method handling the MIT case.

**Pull request merged**